consensus/hotstuff/safetyrules/safety_rules.go

```diff
@@ -93,16 +118,86 @@ func (v *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*mo
 }
 
 // IsSafeToVote checks if this proposal is valid in terms of voting rules, if voting for this proposal won't break safety rules.
-func (v *SafetyRules) IsSafeToVote(proposal *model.Proposal) bool {
-	panic("to be implemented")
+func (r *SafetyRules) IsSafeToVote(proposal *model.Proposal) error {
+	blockView := proposal.Block.View
+	qcView := proposal.Block.QC.View
+
+	// block's view must be larger than the view of the included QC
+	if blockView <= qcView {
+		return fmt.Errorf("block's view %d must be larger than the view of the included QC %d", blockView, qcView)
+	}
+
+	// This check satisfies voting rule 1
+	// 1. Replicas vote strictly in increasing rounds,
+	// block's view must be greater than the view that we have voted for
+	if blockView <= r.safetyData.HighestAcknowledgedView {
+		return model.NewNoVoteErrorf("not safe to vote, we have already voted for this view: %d <= %d",
+			blockView, r.safetyData.HighestAcknowledgedView)
+	}
+
+	// This check satisfies voting rule 2a:
+	// 2a. [Happy path] If the previous round resulted in a QC then new QC should extend it.
+	if blockView == qcView+1 {
+		return nil
+	}
+
+	return r.IsSafeToExtend(blockView, qcView, proposal.LastViewTC)
+}
+
+// IsSafeToExtend performs safety checks if proposal can be extended in case of recovery path, we will call
+// this function only if previous round resulted in TC, to know if it's safe to extend such proposal.
+func (r *SafetyRules) IsSafeToExtend(blockView, qcView uint64, lastViewTC *flow.TimeoutCertificate) error {
+	// These checks satisfy voting rule 2b:
+	// [Recovery Path] If the previous round did *not* result in a QC, the leader of the
+	// subsequent round *must* include a valid TC for the previous round in its block.
+	if lastViewTC == nil {
+		return fmt.Errorf("block's view %d is not sequential with included QC view %d, last view TC not included", blockView, qcView)
+	}
+	if blockView != lastViewTC.View+1 {
+		return fmt.Errorf("last view TC %d is not sequential for block %d", lastViewTC.View, blockView)
+	}
+	if qcView < lastViewTC.TOHighestQC.View {
+		return fmt.Errorf("QC's view %d should be at least %d", qcView, lastViewTC.TOHighestQC.View)
+	}
+	return nil
 }
 
 // IsSafeToTimeout checks if it's safe to timeout with proposed data, if timing out won't break safety rules.
-func (v *SafetyRules) IsSafeToTimeout(curView uint64, highestQC *flow.QuorumCertificate, highestTC *flow.TimeoutCertificate) bool {
-	panic("to be implemented")
+// highestQC is the valid QC with the greatest view that we have observed.
+// lastViewTC is the TC for the previous view (might be nil)
+func (r *SafetyRules) IsSafeToTimeout(curView uint64, highestQC *flow.QuorumCertificate, lastViewTC *flow.TimeoutCertificate) bool {
+	if highestQC.View < r.safetyData.LockedOneChainView ||
+		curView+1 <= r.safetyData.HighestAcknowledgedView ||
+		curView <= highestQC.View {
+		return false
+	}
+
+	return (curView == highestQC.View+1) || (curView == lastViewTC.View+1)
```

Comment:
It would be a bug if we rely on very specific implementation of modules which provide `newestQC`, `curView` and `lastViewTC`. Don't have strong feelings here.